File: Part 2 - Regression/Section 6 - Polynomial Regression/polynomial_regression.R

# ====> Import dataset
setwd("~/work/mlatoz/Part 2 - Regression/Section 6 - Polynomial Regression")
library(ggplot2)
data = read.csv('Position_Salaries.csv')
dataset = data[2:3]

# ====> training set and test set
# library(caTools)
# set.seed(123)
# split = sample.split(dataset$Salary, SplitRatio = 0.8)
# training_set = subset(dataset, split == TRUE)
# test_set = subset(dataset, split == FALSE)

# ====> Linear regression
linear_reg = lm(data = dataset, formula = Salary ~ .)
summary(linear_reg)

# ====> Polynomial regression
dataset$Level2 = dataset$Level^2
dataset$Level3 = dataset$Level^3
polynominal_reg = lm(formula = Salary ~ ., data = dataset)
summary(polynominal_reg)

# ====> Visualising the Linear graph
linear_plot = ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary), color = 'red') +
  geom_line(aes(x = dataset$Level, y = predict(linear_reg, newdata = dataset)), color = 'blue') +
  ggtitle('Truth or Bluff (Linear Regression) R') +
  xlab('Level') +
  ylab('Salary')

ggsave(filename = 'R_linear_regression.png', plot=linear_plot)
  
# ====> Visualising the polynomial graph
polynomial_plot = ggplot() +
  geom_point(aes(x= dataset$Level, y= dataset$Salary), color='red') +
  geom_line(aes(x= dataset$Level, y=predict(polynominal_reg, newdata = dataset)), color='blue') +
  ggtitle('Truth or Bluff (Polynomial Regression) R') +
  xlab('Level') + 
  ylab('Salary')

ggsave(filename = 'R_polynomial_regression.png', plot = polynomial_plot)


# ====> with degree 4
dataset$Level4 = data$Level^4
polynomial_plot = ggplot() +
  geom_point(aes(x= dataset$Level, y= dataset$Salary), color='red') +
  geom_line(aes(x= dataset$Level, y=predict(polynominal_reg, newdata = dataset)), color='blue') +
  ggtitle('Truth or Bluff (Polynomial Regression) R') +
  xlab('Level') + 
  ylab('Salary')

ggsave(filename = 'R_polynomial_regression.png', plot = polynomial_plot)



# =====> Linear regression predict
y_pred = predict(linear_reg, newdata = data.frame(Level = 6.5))
print(y_pred)

# ====> Polynomial regression predict
y_poly_pred = predict(polynominal_reg, newdata = data.frame(Level = 6.5, Level2 = 6.5^2, Level3 = 6.5^3, Level4 = 6.5^4))
print(y_poly_pred)
# ====> Regression model with high resoultion

# level_grid = seq(min(dataset$Level), max(dataset$Level), by=0.1)
# 
# ggplot() +
#   geom_point(aes(x= dataset$Level, y= dataset$Salary), color='red') +
#   geom_line(aes(x= level_grid, y=predict(polynominal_reg, newdata = data.frame(Level = level_grid))), color='blue') +
#   ggtitle('Truth or Bluff (Polynomial Regression) R') +
#   xlab('Level') + 
#   ylab('Salary')
# 

File: Part 2 - Regression/Section 6 - Polynomial Regression/polynomial_regression.py

# ====> importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures

# Importing the data set
dataset = pd.read_csv('Position_Salaries.csv')
level = dataset.iloc[:, 1:2].values
salary = dataset.iloc[:, 2].values

# ====> splitting data into training and data set not required as we are considering full data set for trianing.

# ====> Linear regression
linear_regressor = LinearRegression()
linear_regressor.fit(level, salary)


# =====> Fitting Polynomial regression to the data set.
polynominal_regressor = PolynomialFeatures(degree=4)
level_poly = polynominal_regressor.fit_transform(level)
linear_reg = LinearRegression()
linear_reg.fit(level_poly, salary)


# ====> plot linear regression graph
plt.scatter(level, salary, color='red')
plt.plot(level, linear_regressor.predict(level), color='blue')
plt.title('Truth or Bluff (Linear Regression)')
plt.xlabel('Position Level') 
plt.ylabel('Salary')
plt.savefig('linear_regression.png')
plt.close()

# ====> Polynomial regression model graph
plt.scatter(level, salary, color='red')
plt.plot(level, linear_reg.predict(polynominal_regressor.fit_transform(level)), color='blue')
plt.title('Truth or Bluff (Polynomial Regression)')
plt.xlabel('Position Level') 
plt.ylabel('Salary')
plt.savefig('polynomial_regression.png')
plt.close()

# =====> Polynominal regression model with degree
polynominal_regressor = PolynomialFeatures(degree=3)
level_poly = polynominal_regressor.fit_transform(level)
linear_reg = LinearRegression()
linear_reg.fit(level_poly, salary)

# ====> polynomial regression with degree 3 graph and for higher resolution
level_grid = np.arange(min(level), max(level), 0.1)
level_grid = level_grid.reshape(len(level_grid), 1)
plt.scatter(level, salary, color='red')
plt.plot(level_grid, linear_reg.predict(polynominal_regressor.fit_transform(level_grid)), color='blue')
plt.title('Truth or Bluff (Polynomial Regression)')
plt.xlabel('Position Level') 
plt.ylabel('Salary')
plt.savefig('polynomial_3d_regression.png')
plt.close()


# ====> Linear regression prediction 
linear_regressor.predict(6.5)
# ====> Polynominal regression prediction
linear_reg.predict(polynominal_regressor.fit_transform(6.5))